docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,616 @@
+no one else should read or write from it.
+If no goroutines are writing to the resource but all are merely reading from it,
+that is acceptable, since all goroutines read the same information.
+
+### 4.1.1 Contentions with reading
+
+Read operations are carried out through an iterator, which only reads one chunk of telemetry
+data at a time.
+This means that every time the caller reads a block of telemetry data,
+it first gets the information about the location and length of the data in the file system,
+then uses that information to read on disk.
+
+With this approach, the caller must understand that if the index was changed between two domains –
+say, the iterator was on position 1, and on a call to `Next()`, it moves on to position 2.
+However, a writer inserts a new domain between the two, meaning that the iterator thinks it is on position 2,
+when in reality it loaded the domain at position 3.
+
+Other than this race condition with the index that the user must beware of,
+read does not conflict with any other operations as it does not change the index nor the underlying data.
+
+### 4.1.2 Write-Delete contention
+
+Deletion works by rewriting the pointers where the specified time range's start and end
+time stamps are found, and removing all pointers in between.
+Deletion itself executes no disk operations on telemetry samples in the file system. Instead,
+it accomplishes the 'deletion' by directly changing the index.
+so we only need to consider the concurrent operations on the index by the writer and the
+deleter. We will discuss the two cases: either the writer may write data to the range
+requested to be deleted, or it may not.
+
+In the case that the writer may not write to the deleted time range, the mutex on the index
+is sufficient to handle the concurrent writes as the writer and deleter are necessarily
+writing to different parts of the index, so there will be no conflicts.
+
+The case where the writer may write to the deleted time range, however, will cause index
+conflicts: if the write operation happens first, then a new domain will be inserted and
+it will be deleted from the deletion operation. However, if the deletion operation happens
+first, then there will be a chunk of written data in the time range despite the delete call.
+This is not the behaviour we want. Since we cannot know, if called concurrently, whether
+the delete happened first or the write, we use the controller to disallow all deletes
+to a time range that a writer may write to.
+
+### 4.1.3 Write-GC Contention
+
+Both the writer and the garbage collector modify the underlying data as well as the index.
+In addition, since garbage collection is fundamentally a secondary operation (compared
+to reads, writes, and deletes), it should not affect those operations – it is unrealistic
+to disallow reads and writes while garbage collecting.
+
+there are major conflicts between a writer and GC: Garbage Collection is strictly only
```

Comment:
Capitalize. In addition, I would not abbreviate garbage collection as GC, I would just say garbage collection / collector the whole time - abbreviations can hinder communication more than help. As a rule of thumb, I avoid all abbreviations that aren't common / well known to the audience (e.g. GUI is a good abbreviation, GC is a bad abbreviation).